### PR TITLE
ENT-296: Enable account-level consent requirement

### PR DESCRIPTION
**Pull request by haikuginger**
**Description:**

At present, users are prompted to consent to the exact scope of consent that's relevant to the consent gate that they're attempting to pass - for example, if attempting to access a consent-gated courses, they'll be prompted for course consent; if attempting to access a consent-gated SSO provider, they'll be prompted for account consent.

This change adds a feature whereby, on a per-Enterprise Customer basis, a setting can be made that will force users to be prompted for account-level consent rather than course-specific consent. This change also patches the `EnterpriseCourseEnrollment` save method such that when a record is being saved for an EnterpriseCustomer with the above flag set, an account-wide consent audit record is written.

**JIRA:** [ENT-296](https://openedx.atlassian.net/browse/ENT-296)

**Dependencies:** None

**Sandbox:** https://ent296.opencraft.hosting/

**Merge deadline:** 18 June 2017

**Installation instructions:** No special installation steps; standard Enterprise installation

**Testing instructions:**

1. In [the Enterprise admin panel](https://ent296.opencraft.hosting/admin/enterprise/enterprisecustomer/fcd2b0e7-c687-4e19-af64-8889d835c15f/manage_learners), proxy-enroll a user in two courses
2. Ensure that the [EnterpriseCustomer is set to](https://ent296.opencraft.hosting/admin/enterprise/enterprisecustomer/fcd2b0e7-c687-4e19-af64-8889d835c15f/)
    1. require consent at enrollment
    2. *not* require account-level consent
3. As a user, browse to [one of](https://ent296.opencraft.hosting/courses/course-v1:OpenCraft+Secondary+Run1/courseware) [the courses](https://ent296.opencraft.hosting/courses/course-v1:edX+DemoX+Demo_Course/courseware) that you were enrolled in.
4. Observe the consent prompt; note that the messaging is specific to the single course you're attempting to access. Do not provide consent at this time.
5. Ensure that [EnterpriseCustomer is set to](https://ent296.opencraft.hosting/admin/enterprise/enterprisecustomer/fcd2b0e7-c687-4e19-af64-8889d835c15f/) require account-level consent.
6. Again, browse to one of the courses.
7. Observe that the messaging has changed, and is distinctly supportive of the premise that consent will be given for the entire account.
8. Provide consent.
9. Navigate to the dashboard and click on the _other_ course; note that you're taken directly to it without being prompted for consent.
10. In the Enterprise admin panel for the Enterprise Customer User; note that an account-wide consent audit log has been written, indicating global consent has been provided.

**Merge checklist:**

- [x] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=127&repo=edx%2Fedx-enterprise"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by haikuginger** [comment]
@itsjeyd, this is all set with a sandbox for you to test on.

**Comment by itsjeyd** [comment]
Thanks @haikuginger, I'm going to start looking at this now.

File: enterprise/views.py [hunk 1]
```diff
@@ -284,6 +284,24 @@ def get_course_specific_consent(self, request, course_id):
...
+                ).format(
+                    enterprise_customer_name=customer.name,
+                    bold_start='<b>',
+                    bold_end='</b>',
+                ),
+                'requested_permissions': [
+                    _('your enrollment in all sponsored courses'),
+                    _('your learning progress'),
+                    _('course completion'),
+                ],
```
Comment (itsjeyd) [inline enterprise/views.py]: @haikuginger This list appears in three different places in this file. Wondering if there is a sensible way to define it once, and then just reference as necessary?

File: enterprise/models.py [hunk 2]
```diff
@@ -813,6 +821,34 @@ def __repr__(self):
...
+            # If explicit consent has been provided, create a record indicating that.
+            account_consent_equivalent = UserDataSharingConsentAudit.ENABLED
+        elif self.consent_granted is not None:
+            # If consent was declined, create a record indicating that.
+            account_consent_equivalent = UserDataSharingConsentAudit.DISABLED
+        else:
+            # If no explicit consent state was set, indicate that.
+            account_consent_equivalent = UserDataSharingConsentAudit.NOT_SET
+
+        # This is an awful hack. Once we migrate to the planned generic consent architecture, it can DIAF.
```
Comment (itsjeyd) [inline enterprise/models.py]: @haikuginger Out of curiosity, how are we going to replace this hack once the generic consent architecture lands?

File: tests/test_models.py [hunk 3]
```diff
@@ -150,6 +150,26 @@ def test_consent_not_stored_audit_available_externally_managed(self):
...
+    @ddt.data(
+        ('at_login', None, 1, 'not_set'),
+        ('at_login', True, 1, 'enabled'),
+        ('at_login', False, 1, 'disabled'),
+        ('externally_managed', None, 1, 'external'),
+        ('externally_managed', False, 1, 'external'),
+        ('externally_managed', True, 1, 'external'),
+    )
+    @ddt.unpack
+    def test_create_related_consent_audit(self, requirement, this_state, expected_count, expected_value):
```
Comment (itsjeyd) [inline tests/test_models.py]: @haikuginger Looks like `expected_count` is always `1` -- can we get rid of this parameter (and the `if` in L170 below)?

File: tests/test_views.py [hunk 4]
```diff
@@ -524,6 +524,93 @@ def test_get_course_specific_consent(
...
+        (False, False),
+        (True, True),
+    )
+    @ddt.unpack
+    def test_get_course_specific_consent_ec_requires_account_level(
+            self,
+            enrollment_deferred,
+            supply_customer_uuid,
+            course_api_client_mock,
+            render_mock,  # pylint: disable=unused-argument
```
Comment (itsjeyd) [inline tests/test_views.py]: @haikuginger Nit: [`render_mock_unused`](https://github.com/edx/edx-enterprise/blob/master/pylintrc#L155)? Then we can remove `# pylint: disable=unused-argument`.

**Comment by itsjeyd** [comment]
@haikuginger @brittneyexline Getting a 500 for my user as well. Can access both courses just fine when logged in as `staff`.

File: enterprise/views.py (hunk 1 above)
Comment (haikuginger) [inline enterprise/views.py]: Two of the instances are the same; the other is different. If they were all the same, then I absolutely would, haha.

File: enterprise/models.py (hunk 2 above)
Comment (haikuginger) [inline enterprise/models.py]: We'll consolidate to a single consent model that can have different scopes stored natively.

File: tests/test_models.py (hunk 3 above)
Comment (haikuginger) [inline tests/test_models.py]: Sure!

File: tests/test_views.py (hunk 4 above)
Comment (haikuginger) [inline tests/test_views.py]: Sure!

**Comment by haikuginger** [comment]
@itsjeyd @brittneyexline ready for a second pass. Fixed the configuration issue preventing the sandbox from working correctly.

File: enterprise/views.py (hunk 1 above)
Comment (itsjeyd) [inline enterprise/views.py]: @haikuginger Haha, OK :) I could have sworn they were identical :P

**Comment by itsjeyd** [comment]
@haikuginger Thanks for the updates! The code looks good, giving this a spin on the sandbox now.

**Pull request merged**